packages/ui-extensions-sdk/src/utils/security.ts

```diff
@@ -1,19 +1,41 @@
 import { LoadedResourceMetaDataBatch, NetworkRequestMetadata } from '../types';
 
-// TODO: Feel free to pass in whatever data structure you need
-// to track what has been previously loaded. Here I'm assuming it will be a set of
-// string ids
 export type LoadedResourceIds = Set<string>;
 
+function isRequestKind(timing: PerformanceResourceTiming) {
+    return (
+        timing.initiatorType === 'xmlhttprequest' ||
+        timing.initiatorType === 'fetch'
+    );
+}
+
 export const collectResourceUsage = (
     previouslyLoadedResources: LoadedResourceIds
 ): [LoadedResourceMetaDataBatch, LoadedResourceIds] => {
-    const newLoadedResourceIds = new Set(previouslyLoadedResources);
+    // Use resource timing api to to retrieve application loaded resources
+    const snapshot = performance.getEntriesByType('resource');
 
-    // TODO: Do batch collection of resource data
-    // TODO: Update list of preivously loaded resource ids
+    // to avoid collecting the same resource twice, we use
+    // the Set difference: snapshot - previouslyLoadedResources
+    const resources = snapshot
+        .filter(r => !previouslyLoadedResources.has(r.name))
+        .map(e => {
+            const a = e as PerformanceResourceTiming;
+            const urlHostname = new URL(a.name).hostname;
+            let url;
+            if (!isRequestKind(a)) {
+                // for privacy reason we do not capture full URLs of network calls
+                url = a.name;
+            }
+            return {
+                url,
+                urlHostname,
+                initiatorType: a.initiatorType,
+                nextHopProtocol: a.nextHopProtocol
+            };
+        });
 
-    return [{ resources: [] }, newLoadedResourceIds];
+    return [{ resources }, new Set(snapshot.map(x => x.name))];
```

Comment:
concern: You are setting the 'loadedResourceIds' to be the names of the current `snapshot`, but don't you want to accumulate _all_ of the ids you've ever loaded?

```
const updatedIds = new Set(previouslyLoadedResources);

snapshot.forEach(item => { updatedIds.add(item.name) })
```

Reply:
You're right I'm setting the `loadedResourceIds` from the current snapshot instead of accumulating all previously loaded resources.

I think deduplication would still work with that approach because
- before `resourcetimingbufferfull` triggers : `previouslyLoadedResources` is always subset of `performance.getEntriesByType('resource')`, so setting the `loadedResourceIds` to current snapshot should be exactly the same as accumulating all previous IDs.
- If `resourcetimingbufferfull` triggers : performances entries buffer will be reset and `collectResources` will not report resources that have already been reported so we have no risk of sending duplicate resources.

This approach will avoid having an unbounded `loadedResourceIds` Set, it will be reseted when `resourcetimingbufferfull` is reached.

But maybe this is misleading/obscure given the name `loadedResourceIds` which might indicates it contains all previous ids. Do you think renaming var `loadedResourceIds` to something like `previousSnapshotResourceIds` would help ?

Reply:
Got it, thanks for the explanation!